tree/ntuple/v7/inc/ROOT/RNTupleProcessor.hxx

```diff
@@ -0,0 +1,287 @@
+   /// \brief Returns a reference to the page source currently used in the processor
+   ///
+   /// \return A reference to the page source
+   Internal::RPageSource &GetPageSource() const { return *fPageSource; }
+
+public:
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Returns a reference to the entry used by the processor.
+   ///
+   /// \return A reference to the entry used by the processor.
+   ///
+   const REntry &GetEntry() const { return *fEntry; }
+
+   // clang-format off
+   /**
+   \class ROOT::Experimental::RNTupleProcessor::RIterator
+   \ingroup NTuple
+   \brief Iterator over the entries of an RNTuple, or vertical concatenation thereof.
+   */
+   // clang-format on
+   class RIterator {
+   private:
+      RNTupleProcessor &fProcessor;
+      const std::vector<RNTupleSourceSpec> fNTuples;
+      std::size_t fNTupleIndex;
+      NTupleSize_t fGlobalEntryIndex;
+      NTupleSize_t fLocalEntryIndex;
+      RProcessorView fProcessorView;
+
+   public:
+      using iterator_category = std::forward_iterator_tag;
+      using iterator = RIterator;
+      using value_type = RProcessorView;
+      using difference_type = std::ptrdiff_t;
+      using pointer = RProcessorView *;
+      using reference = const RProcessorView &;
+
+      RIterator(RNTupleProcessor &processor, const std::vector<RNTupleSourceSpec> &ntuples, std::size_t ntupleIndex,
+                NTupleSize_t globalEntryIndex)
+         : fProcessor(processor),
+           fNTuples(ntuples),
+           fNTupleIndex(ntupleIndex),
+           fGlobalEntryIndex(globalEntryIndex),
+           fLocalEntryIndex(0),
+           fProcessorView(processor.GetEntry(), fGlobalEntryIndex, fLocalEntryIndex, fNTupleIndex)
+      {
+      }
+
+      //////////////////////////////////////////////////////////////////////////
+      /// \brief Increments the entry index.
+      ///
+      /// Checks if the end of the currently connected RNTuple is reached. If this is the case, either the next RNTuple
+      /// is connected or the iterator has reached the end.
+      void Advance()
+      {
+         ++fGlobalEntryIndex;
+
+         if (++fLocalEntryIndex >= fProcessor.GetPageSource().GetNEntries()) {
+            do {
+               if (++fNTupleIndex >= fNTuples.size()) {
+                  fGlobalEntryIndex = kInvalidNTupleIndex;
+                  return;
+               }
+               // Skip over empty ntuples we might encounter.
+            } while (fProcessor.ConnectNTuple(fNTuples.at(fNTupleIndex)) == 0);
+
+            fLocalEntryIndex = 0;
+         }
+      }
+
+      iterator operator++()
+      {
+         Advance();
+         return *this;
+      }
+
+      iterator operator++(int)
+      {
+         auto obj = *this;
+         Advance();
+         return obj;
+      }
+
+      reference operator*()
+      {
+         if (fProcessor.GetPageSource().GetNEntries() == 0)
+            Advance();
```

Comment:
Can this happen since `Advance()` already takes care of skipping empty RNTuples. Perhaps what's missing is an initial check for empty RNTuple(s) on construction. We can test if the processor works with one or a set of empty files.